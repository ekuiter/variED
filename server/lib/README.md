## Libraries

This folder contains the FeatureIDE core library
(`plugins/de.ovgu.featureide.fm.core`) and its dependencies.

The core library (`de.ovgu.featureide.lib.fm.jar`) was built from [this
commit](https://github.com/FeatureIDE/FeatureIDE/tree/aec4288772d3a1cfb89a0cd14a9abb6281daedf5/plugins/de.ovgu.featureide.fm.core/library).

The additional libraries can be found
[here](https://github.com/FeatureIDE/FeatureIDE/tree/aec4288772d3a1cfb89a0cd14a9abb6281daedf5/plugins/de.ovgu.featureide.fm.core/lib).
