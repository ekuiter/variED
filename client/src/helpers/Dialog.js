import {Dialog, DialogFooter} from 'office-ui-fabric-react/lib/Dialog';
import {DefaultButton, PrimaryButton} from 'office-ui-fabric-react/lib/Button';
import i18n from '../i18n';
import React from 'react';
import defer from './defer';
import {TextField} from '../../node_modules/office-ui-fabric-react/lib/TextField';
import PropTypes from 'prop-types';
import exact from 'prop-types-exact';

export const DialogWrapper = ({isOpen, onDismiss, onApply, children, label}) => (
    <Dialog
        hidden={!isOpen}
        onDismiss={onDismiss}
        dialogContentProps={{title: label}}>
        {children}
        <DialogFooter>
            <PrimaryButton onClick={onApply} text={i18n.t('panels.settingsPanel.apply')}/>
        </DialogFooter>
    </Dialog>
);

DialogWrapper.propTypes = exact({
    isOpen: PropTypes.bool.isRequired,
    onDismiss: PropTypes.func.isRequired,
    onApply: PropTypes.func.isRequired,
    children: PropTypes.node.isRequired,
    label: PropTypes.string.isRequired
});

export class DialogContextualMenu extends React.Component {
    static propTypes = exact({
        onApply: PropTypes.func.isRequired,
        onRender: PropTypes.func.isRequired,
        options: PropTypes.arrayOf(PropTypes.object).isRequired,
        children: PropTypes.node.isRequired,
        label: PropTypes.string.isRequired,
        iconProps: PropTypes.object
    });

    state = {option: null};
    onDialogDismiss = () => this.setState({option: null});

    onDialogOpen = option => () => {
        this.props.onRender(option);
        this.setState({option});
    };

    onApply = () => {
        this.props.onApply(this.state.option);
        this.onDialogDismiss();
    };

    render() {
        return (
            <div className="setting">
                <DefaultButton
                    text={this.props.label}
                    iconProps={this.props.iconProps}
                    menuProps={{
                        items: this.props.options.map(option =>
                            ({...option, onClick: this.onDialogOpen(option)}))
                    }}/>
                {this.state.option &&
                <DialogWrapper
                    label={this.state.option.text}
                    isOpen={!!this.state.option}
                    onDismiss={this.onDialogDismiss}
                    onApply={this.onApply}>
                    {this.props.children}
                </DialogWrapper>}
            </div>
        );
    }
}

export class TextFieldDialog extends React.Component {
    static propTypes = {
        isOpen: PropTypes.bool.isRequired,
        onDismiss: PropTypes.func.isRequired,
        onSubmit: PropTypes.func.isRequired,
        title: PropTypes.string.isRequired,
        defaultValue: PropTypes.string,
        submitText: PropTypes.string.isRequired,
        submitOnEnter: PropTypes.bool,
        textFieldProps: PropTypes.object
    };

    static defaultProps = {submitOnEnter: true};
    state = {value: null};
    textFieldRef = React.createRef();
    onChange = (e, value) => this.setState({value});

    // does not work without defer, god knows why
    onLayerDidMount = defer(() => {
        this.textFieldRef.current.focus();
        this.textFieldRef.current.select();
    });

    getValue = () => this.state.value === null ? this.props.defaultValue || '' : this.state.value;

    onSubmit = () => {
        this.props.onSubmit(this.getValue());
        this.setState({value: null});
        this.props.onDismiss();
    };

    onKeyPress = e => {
        if (e.key === 'Enter' && this.props.submitOnEnter)
            this.onSubmit();
    };

    render() {
        const {isOpen, onDismiss, onSubmit: _onSubmit, title, defaultValue: _defaultValue,
            submitText, submitOnEnter: _submitOnEnter, textFieldProps, ...props} = this.props;
        return (
            <Dialog
                hidden={!isOpen}
                onDismiss={onDismiss}
                modalProps={{onLayerDidMount: this.onLayerDidMount}}
                dialogContentProps={{title}}
                {...props}>
                <TextField
                    componentRef={this.textFieldRef}
                    value={this.getValue()}
                    onChange={this.onChange}
                    onKeyPress={this.onKeyPress}
                    {...textFieldProps}/>
                <DialogFooter>
                    <PrimaryButton onClick={this.onSubmit} text={submitText}/>
                </DialogFooter>
            </Dialog>
        );
    }
}