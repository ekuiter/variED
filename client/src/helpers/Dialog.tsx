import {Dialog, DialogFooter} from 'office-ui-fabric-react/lib/Dialog';
import {DefaultButton, PrimaryButton} from 'office-ui-fabric-react/lib/Button';
import i18n from '../i18n';
import React from 'react';
import defer from './defer';
import {TextField, ITextFieldProps, ITextField} from 'office-ui-fabric-react/lib/TextField';
import {IContextualMenuItem} from 'office-ui-fabric-react/lib/ContextualMenu';
import {IIconProps} from 'office-ui-fabric-react/lib/Icon';

export const largeDialogStyle = {
    main: {
        selectors: {
            '@media (min-width: 480px)': {minWidth: 400, maxWidth: 500},
            '@media (min-width: 720px)': {minWidth: 500, maxWidth: 600}
        }
    }
};

export const DialogWrapper = ({isOpen, onDismiss, onApply, children, label}:
    {isOpen: boolean, onDismiss: () => void, onApply: () => void, children: JSX.Element[], label: string}) => (
    <Dialog
        hidden={!isOpen}
        onDismiss={onDismiss}
        dialogContentProps={{title: label}}>
        {children}
        <DialogFooter>
            <PrimaryButton onClick={onApply} text={i18n.t('overlays.settingsPanel.apply')}/>
        </DialogFooter>
    </Dialog>
);

interface DialogContextualMenuProps {
    onApply: (option: IContextualMenuItem) => void,
    onRender: (option: IContextualMenuItem) => void,
    options: IContextualMenuItem[],
    children: JSX.Element[],
    label: string
    iconProps?: IIconProps
};

interface DialogContextualMenuState {
    option?: IContextualMenuItem
};

export class DialogContextualMenu extends React.Component<DialogContextualMenuProps, DialogContextualMenuState> {
    state: DialogContextualMenuState = {option: undefined};
    onDialogDismiss = () => this.setState({option: undefined});

    onDialogOpen = (option: IContextualMenuItem) => () => {
        this.props.onRender(option);
        this.setState({option});
    };

    onApply = () => {
        this.props.onApply(this.state.option!);
        this.onDialogDismiss();
    };

    render() {
        return (
            <div className="setting">
                <DefaultButton
                    text={this.props.label}
                    iconProps={this.props.iconProps}
                    menuProps={{
                        items: this.props.options.map(option =>
                            ({...option, onClick: this.onDialogOpen(option)}))
                    }}/>
                {this.state.option &&
                <DialogWrapper
                    label={this.state.option!.text!}
                    isOpen={!!this.state.option}
                    onDismiss={this.onDialogDismiss}
                    onApply={this.onApply}>
                    {this.props.children}
                </DialogWrapper>}
            </div>
        );
    }
}

interface TextFieldDialogProps {
    isOpen: boolean,
    onDismiss: () => void,
    onSubmit: (value: string) => void,
    title: string,
    defaultValue?: string,
    submitText: string,
    submitOnEnter?: boolean,
    textFieldProps?: ITextFieldProps
};

interface TextFieldDialogState {
    value?: string
};

export class TextFieldDialog extends React.Component<TextFieldDialogProps, TextFieldDialogState> {
    static defaultProps: Partial<TextFieldDialogProps> = {submitOnEnter: true};
    state: TextFieldDialogState = {value: undefined};
    textFieldRef = React.createRef<ITextField>();
    onChange = (_event: any, value?: string) => this.setState({value});

    // does not work without defer, god knows why
    onLayerDidMount = defer(() => {
        this.textFieldRef.current!.focus();
        this.textFieldRef.current!.select();
    });

    getValue = (): string => typeof this.state.value === 'undefined' ? this.props.defaultValue || '' : this.state.value;

    onSubmit = () => {
        this.props.onSubmit(this.getValue());
        this.setState({value: undefined});
        this.props.onDismiss();
    };

    onKeyPress = (event: React.KeyboardEvent) => {
        if (event.key === 'Enter' && this.props.submitOnEnter)
            this.onSubmit();
    };

    render() {
        const {isOpen, onDismiss, onSubmit: _onSubmit, title, defaultValue: _defaultValue,
            submitText, submitOnEnter: _submitOnEnter, textFieldProps, ...props} = this.props;
        return (
            <Dialog
                hidden={!isOpen}
                onDismiss={onDismiss}
                modalProps={{onLayerDidMount: this.onLayerDidMount}}
                dialogContentProps={{title}}
                {...props}>
                <TextField
                    componentRef={this.textFieldRef}
                    value={this.getValue()}
                    onChange={this.onChange}
                    onKeyPress={this.onKeyPress}
                    {...textFieldProps}/>
                <DialogFooter>
                    <PrimaryButton onClick={this.onSubmit} text={submitText}/>
                </DialogFooter>
            </Dialog>
        );
    }
}