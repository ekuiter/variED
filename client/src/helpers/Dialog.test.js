import React from 'react';
import {shallow, mount} from 'enzyme';
import {DialogContextualMenu, TextFieldDialog, DialogWrapper} from './Dialog';
import {DefaultButton, PrimaryButton} from 'office-ui-fabric-react/lib/Button';
import {Dialog} from 'office-ui-fabric-react/lib/Dialog';
import {TextField} from 'office-ui-fabric-react/lib/TextField';

describe('Dialog', () => {
    describe('DialogWrapper', () => {
        it('renders correctly', () => {
            const mock = jest.fn(),
                wrapper = shallow(
                    <DialogWrapper
                        label="label"
                        isOpen={true}
                        onDismiss={mock}
                        onApply={mock}>
                    content
                    </DialogWrapper>
                );
            expect(wrapper.prop('dialogContentProps').title).toBe('label');
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('DialogContextualMenu', () => {
        const dialogContextualMenu = (onApply = jest.fn(), onRender = jest.fn()) =>
            shallow(
                <DialogContextualMenu
                    label="label"
                    options={[{key: 'key', text: 'text'}]}
                    onApply={onApply}
                    onRender={onRender}>
                content
                </DialogContextualMenu>
            );

        it('renders correctly', () => {
            const wrapper = dialogContextualMenu();
            expect(wrapper.find(DefaultButton).prop('text')).toBe('label');
            expect(wrapper.find(DefaultButton).prop('menuProps').items[0].key).toBe('key');
            expect(wrapper.contains('content')).not.toBe(true);
            expect(wrapper).toMatchSnapshot();
        });

        it('opens a dialog when clicking an option', () => {
            const onApply = jest.fn(),
                onRender = jest.fn(),
                wrapper = dialogContextualMenu(onApply, onRender);
            wrapper.find(DefaultButton).prop('menuProps').items[0].onClick();
            expect(onRender).toBeCalledWith({key: 'key', text: 'text'});
        });

        it('closes the dialog when it is applied', () => {
            const onApply = jest.fn(),
                onRender = jest.fn(),
                wrapper = dialogContextualMenu(onApply, onRender),
                dialogWrapper = wrapper => wrapper.find(DialogWrapper);
            expect(dialogWrapper(wrapper).get(0)).toBeFalsy();
            wrapper.find(DefaultButton).prop('menuProps').items[0].onClick();
            expect(dialogWrapper(wrapper).get(0)).toBeTruthy();
            dialogWrapper(wrapper).simulate('apply');
            expect(onApply).toBeCalledWith({key: 'key', text: 'text'});
            expect(dialogWrapper(wrapper).get(0)).toBeFalsy();
        });
    });

    describe('TextFieldDialog', () => {
        const textFieldDialog = (onSubmit = jest.fn(), onDismiss = jest.fn(), isOpen = true, render = shallow) =>
            render(
                <TextFieldDialog
                    isOpen={isOpen}
                    title="title"
                    submitText="submitText"
                    onSubmit={onSubmit}
                    onDismiss={onDismiss}/>
            );

        it('renders correctly', () => {
            const wrapper = textFieldDialog();
            expect(wrapper.find(Dialog).prop('dialogContentProps').title).toBe('title');
            expect(wrapper.find(PrimaryButton).prop('text')).toBe('submitText');
            expect(wrapper).toMatchSnapshot();
        });

        it('focuses a text when it opens', () => {
            jest.useFakeTimers();
            const wrapper = textFieldDialog(jest.fn(), jest.fn(), false, mount);
            const onLayerDidMount = jest.spyOn(wrapper.instance(), 'onLayerDidMount');
            expect(onLayerDidMount).not.toBeCalled();
            wrapper.setProps({isOpen: true});
            expect(onLayerDidMount).toBeCalled();
            
            const focus = jest.spyOn(wrapper.instance().textFieldRef.current, 'focus'),
                select = jest.spyOn(wrapper.instance().textFieldRef.current, 'focus');
            expect(focus).not.toBeCalled();
            expect(select).not.toBeCalled();
            wrapper.instance().onLayerDidMount();
            jest.runAllTimers();
            expect(focus).toBeCalled();
            expect(select).toBeCalled();
        });

        it('submits a text', () => {
            const onSubmit = jest.fn(),
                onDismiss = jest.fn(),
                wrapper = textFieldDialog(onSubmit, onDismiss);
            wrapper.find(TextField).simulate('change', null, 'some value');
            wrapper.find(PrimaryButton).simulate('click');
            expect(onSubmit).toBeCalledWith('some value');
            expect(onDismiss).toBeCalled();
        });

        it('submits a text on enter when enabled', () => {
            const onSubmit = jest.fn(),
                onDismiss = jest.fn(),
                wrapper = textFieldDialog(onSubmit, onDismiss);
            wrapper.find(TextField).simulate('change', null, 'some value');
            wrapper.find(TextField).simulate('keyPress', {key: 'Enter'});
            expect(onSubmit).toBeCalledWith('some value');
            expect(onDismiss).toBeCalled();
        });

        it('does not submit a text on enter when disabled', () => {
            const onSubmit = jest.fn(),
                onDismiss = jest.fn(),
                wrapper = textFieldDialog(onSubmit, onDismiss);
            wrapper.setProps({submitOnEnter: false});
            wrapper.find(TextField).simulate('change', null, 'some value');
            wrapper.find(TextField).simulate('keyPress', {key: 'Enter'});
            expect(onSubmit).not.toBeCalled();
            expect(onDismiss).not.toBeCalled();
        });
    });
});