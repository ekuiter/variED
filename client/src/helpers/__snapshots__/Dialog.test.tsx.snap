// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog DialogContextualMenu renders correctly 1`] = `
<div
  className="setting"
>
  <CustomizedDefaultButton
    menuProps={
      Object {
        "items": Array [
          Object {
            "key": "key",
            "onClick": [Function],
            "text": "text",
          },
        ],
      }
    }
    text="label"
  />
</div>
`;

exports[`Dialog DialogWrapper renders correctly 1`] = `
<StyledWithResponsiveMode
  dialogContentProps={
    Object {
      "title": "label",
    }
  }
  hidden={false}
  onDismiss={[Function]}
>
  content
  <StyledDialogFooterBase>
    <CustomizedPrimaryButton
      onClick={[Function]}
      text="Apply"
    />
  </StyledDialogFooterBase>
</StyledWithResponsiveMode>
`;

exports[`Dialog TextFieldDialog renders correctly 1`] = `
<StyledWithResponsiveMode
  dialogContentProps={
    Object {
      "title": "title",
    }
  }
  hidden={false}
  modalProps={
    Object {
      "onLayerDidMount": [Function],
    }
  }
  onDismiss={[Function]}
>
  <StyledTextFieldBase
    componentRef={
      Object {
        "current": null,
      }
    }
    onChange={[Function]}
    onKeyPress={[Function]}
    value=""
  />
  <StyledDialogFooterBase>
    <CustomizedPrimaryButton
      onClick={[Function]}
      text="submitText"
    />
  </StyledDialogFooterBase>
</StyledWithResponsiveMode>
`;
